### PR TITLE
feat(api): pdf generation

### LINKED ISSUE #73: [5JH] Génération de rapport PDF
## Describe the issue
Générer les PDFs lors de la fin de la mission.

## Resource request
- [x]  Finir les CRUDs

## Definition of done
- [ ] Faire un form avec introduction, conclusion
- [ ] (optionnel) demander le logo pour l'inclure dans le pdf
- [x] générer le pdf dans le back-end

## JEH
5

### PR DESCRIPTION
**Pull request by lambdina**
# Description

Added automated report pdf generation

# Changes include
- [x] New feature (non-breaking change that adds functionality)

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added @vrn-sh/members to the reviewers, or specific members of the team
- [x] I have added the needed labels
- [ ] I have tested this code
- [ ] I have added / updated tests (unit / functionals / end-to-end / ...)
- [ ] I have updated the README and other relevant documents (guides...)

# Additional comments

image
image
image
image
image

### CODE REVIEW

File: api/api/services/s3.py [hunk 1]
```diff
@@ -10,7 +10,7 @@
 class S3Bucket:
     def __init__(self) -> None:
         self.client = Minio(
-            "s3:9000",
+            "localhost:9000", #TOFIX
```
Comment (djnnvx) [inline api/api/services/s3.py]: this should be an env value, i think you can pull `fix/s3-image-upload` in order to have the correct code

File: api/api/views/report.py [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+from api.models.vulns import Vulnerability
+from api.services.s3 import S3Bucket
+
+DIR_STYLE = "./api/pdf-templates/hackmanit-template"
+CSS_PATH = f'{DIR_STYLE}/main.css'
+ABSOLUTE_DIR_STYLE = os.path.abspath(DIR_STYLE if os.getenv("PRODUCTION", 0) else DIR_STYLE.replace("api/", ""))
+ABSOLUTE_CSS_PATH = f"{ABSOLUTE_DIR_STYLE}/main.css"
+
+
+def get_image_file_as_base64_data(path):
```
Comment (djnnvx) [inline api/api/views/report.py]: type hints plz

File: api/api/views/report.py [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+    with open(path, 'rb') as image_file:
+        return base64.b64encode(image_file.read())
+
+
+def generate_members(team: Team):
```
Comment (djnnvx) [inline api/api/views/report.py]: please add `-> List[str]:`

File: api/api/views/report.py [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+        mission = Mission.objects.filter(pk=mission_id).first()
+        if not mission:
+            return Response({
+                'error': f'No mission with id {mission_id}. Report couldn\'t be generated',
+            }, status=HTTP_404_NOT_FOUND)
+        s3 = S3Bucket()
+        s3.create_bucket(mission.bucket_name)
+        dir_path = f'/tmp/{mission.bucket_name}'
+        filepath = self.dump_report(mission, dir_path)
+        object_name = f'report-{mission.title}.pdf'
```
Comment (djnnvx) [inline api/api/views/report.py:79]: this is probably vulnerable to LFI, maybe you should generate a random name instead

File: api/api/views/report.py [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+        s3.upload_file(mission.bucket_name, file_path=filepath, file_name=object_name)
+        rmtree(dir_path)
+
+        print(s3.client.presigned_get_object(
```
Comment (djnnvx) [inline api/api/views/report.py]: you can remove this

File: api/api/views/report.py [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+
+    def dump_report(self, mission: Mission, dir_path) -> str:
+        pages = [
+            (f'{dir_path}/coverpage.html', self.generate_cover(mission, mission.team.leader)),
+            (f'{dir_path}/projectinfo.html', self.generate_project_info(mission, mission.team)),
+            (f'{dir_path}/generalconditionsandscope.html', self.generate_condition_and_scope(mission.scope)),
+            (f'{dir_path}/weaknesses.html', self.generate_weaknesses(mission))
+        ]
+
+        print(pages[3][1])
```
Comment (djnnvx) [inline api/api/views/report.py]: you can remove this too

File: api/api/views/report.py [hunk 7]
```diff
@@ -0,0 +1,415 @@
...
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <title>General Conditions and Scope</title>
+    <link rel="stylesheet" type="text/css" href="{stylesheet_path}" />
+</head>
+
+<body>
+    <header>
+        <div class="inline-title-logo">
+            <img alt="logo-company" id="logo"
+                src="https://www.hackmanit.de/templates/hackmanit2021j4/img/wbm_hackmanit.png" />
```
Comment (djnnvx) [inline api/api/views/report.py:269]: we should add the company's logo later. Can you add an issue regarding this ?

File: api/api/views/report.py [hunk 8]
```diff
@@ -0,0 +1,415 @@
...
+            />
+            <p>Figure 2: Successful session initialization with a missing <code>state</code> parameter</p>
+        -->
+        </div>
+            '''.format(vuln_label_class=vuln_label_class,
+                       vuln_label=vuln_label,
+                       vuln_title=vuln.title,
+                       subscore_1=vuln.severity * 0.45, # BS hihi chuuuut
```
Comment (djnnvx) [inline api/api/views/report.py:423]: :o

File: api/core/settings.py [hunk 9]
```diff
@@ -209,7 +209,7 @@
         'default': {
             'ENGINE': 'django.db.backends.postgresql',
             'NAME': 'voron',
-            'USER': os.environ.get('USER'),
+            'USER': 'postgres',
```
Comment (djnnvx) [inline api/core/settings.py]: revert this please. If you need, you can set USER=postgres in your `.env` file

File: api/api/views/report.py [hunk 10]
```diff
@@ -0,0 +1,415 @@
...
+    for member in team.members.all():
+        members_html += f"<p>{member.auth.first_name} {member.auth.last_name}</p>"
+    return members_html
+
+
+class GenerateReportView(APIView):
+    permission_classes = [AllowAny]
+    authentication_classes: List[Type[TokenAuthentication]] = []
+
+    def get(self, request):
```
Comment (djnnvx) [inline api/api/views/report.py:64]: please add API docs for the front-end team

File: api/scripts/setup.sh [hunk 11]
```diff
@@ -32,7 +32,7 @@ install_postgresql () {
 ############
 
 # necessary know dependencies
-sudo apt-get install build-essential libssl-dev libffi-dev python3-dev python-dev
+sudo apt-get install build-essential libssl-dev libffi-dev python3-dev
```
Comment (djnnvx) [inline api/scripts/setup.sh]: you should add `wkhtmltopdf` here, in `setup_mac.sh` and in `Dockerfile`

**Review by djnnvx — CHANGES_REQUESTED** [review]
looks good overall, but a few things to tweak. Let me know if you need help

**Comment by djnnvx** [comment]
seems to be conflicts as well, you should perhaps nuke the migrations and re-generate them. that might be the easiest way since we do not have anything sensible in prod yet

File: api/api/services/s3.py (hunk 1 above)
Comment (lambdina) [reply api/api/services/s3.py]: done

File: api/api/views/report.py (hunk 3 above)
Comment (lambdina) [reply api/api/views/report.py]: done

File: api/api/views/report.py (hunk 5 above)
Comment (lambdina) [reply api/api/views/report.py]: done

File: api/api/views/report.py (hunk 2 above)
Comment (lambdina) [reply api/api/views/report.py]: done

File: api/api/views/report.py (hunk 6 above)
Comment (lambdina) [reply api/api/views/report.py]: done

File: api/api/views/report.py (hunk 4 above)
Comment (lambdina) [reply api/api/views/report.py:79]: won't do for now. Will add an issue.

File: api/core/settings.py (hunk 9 above)
Comment (lambdina) [reply api/core/settings.py]: done

File: api/api/views/report.py (hunk 10 above)
Comment (lambdina) [reply api/api/views/report.py:64]: done

**Review by djnnvx — APPROVED** [review]

**Pull request merged**